### PR TITLE
Use next occurrence for recurring events in search

### PR DESCRIPTION
**Pull request by dcaliste**
A placeholder for a possible implementation of using `getNextOccurrence()` to list recurring events in the serach results. Currently unimplemented.

It initially contains the commit adding the year section.

### CODE REVIEW

**Comment by dcaliste** [comment]
I've added a commit calling `CalendarManager::getNextOccurrence()` for recurring events. This has a potential performence drawback, since such calls are blocking on the worker thread. In my testing I didn't notice any freeze issue though.The worker thread is not asking for storage access neither in that case, but instead just doing some calculation with the memory calendar.

The other problem of the current proposition is a bad interaction with the `limit` property of the search model. Such limit is passed through to mKCal search method, which only loads a restricted by start time list of incidences. The birthday incidences being "far" in the past, they are often not listed if the search is broad and return a lot of incidences. I see two possibilities:
- change the mKCal search method to do the occurrence expansion and apply the limit after. This has the drawback that mKCal is loading then incidences into the calendar, and not occurrences. So these occurrences will have to be recomputed at nemo-qml-plugin-calendar level. And it complicates the search() method API with a QDateTime argument that is not obvious to understand.
- always load recurring and matching incidences in a search call, irrespective of the limit argument. Then let the upstream do the occurrence expansion and reenforce the limit. This is not ideal though since the number of returned incidences by mKCal will be completely arbitrary whatever the limit argument.

In both case, we'll also loose the performence gain induced by the limit : all matching incidences must be loaded from database to check if they are recurring or not and if one of their occurrence is within the first `limit` incidences in the return list...

Besides that, using the current state of this PR works well for "accurate" search and recurring incidences get their closest occurences well positioned in the list without noticeable performence loss.

**Comment by dcaliste** [comment]
I've added a commit on top where the calculation of the next occurrence is now done directly in the main thread and not by a blocking call to the worker thread. This is possible for recurring schemes that are known from Calendar::Event, which are most of the cases. It fallback to the worker thread calculations only for Calendar::RecurCustom cases.

This should solve potential blocking issues, and thread ping-pong, to get the correct occurrences for the search result list.

It does not address the limit issue though and thus recurring events defined after the limit but that have occurrences within are still not displayed.

**Comment by dcaliste** [comment]
With sailfishos/mkcal#58, the occurrence of recurring events can be retrieved, even with the limit argument. So I'm removing the WIP tag.

**Review by pvuorela — COMMENTED** [review]
There seems to be  a unit test regression, failure on tst_calendarevent.cpp(464)

File: src/calendareventlistmodel.cpp [hunk 1]
```diff
@@ -96,9 +96,32 @@ QStringList CalendarEventListModel::missingItems() const
...
 
+// For recurring events, if mStartTime is valid, the stored occurrence
+// is the closest to mStartTime. When mStartTime is invalid, the
+// initial occurrence is used.
+QDateTime CalendarEventListModel::startTime() const
+{
+    return mStartTime;
+}
+
+void CalendarEventListModel::setStartTime(const QDateTime &t)
```
[the remark this answers is not in the document]
Comment (dcaliste) [reply src/calendareventlistmodel.cpp:107]: Ah, yes sorry.

**Comment by dcaliste** [comment]
> There seems to be a unit test regression, failure on tst_calendarevent.cpp(464)

Indeed, my mistake. Doing the occurrence expansion in the main thread is not possible, even for known recurrence scheme, since it's missing the ExDates and the RDates. So I've cut-down the last commit, to still do the getOccurrence thing in the main thread, but only for plain events. For recurring one, it's still back to the worker thread. It's less efficient than the initial thought, but it may still be not bad.

**Comment by pvuorela** [comment]
Works now better, thanks. The worker query on recurring events is unfortunate, but since it's limited to the search I can live with it.

Commented earlier on mkcal on why should we need ordering of recurring events first. Since this does some ordering, could just keep the mkcal side simpler, could be good enough to not even define any ordering for the results? There we also some other changes on mkcal which should be good, though.

**Comment by dcaliste** [comment]
> Commented earlier on mkcal on why should we need ordering of recurring events first. Since this does some ordering, could just keep the mkcal side simpler

That's a valid question. The ordering on mKCal side is to be understand with the limit argument. With the limit argument, mKCal does not need to load all matching incidences. But for the limit to work, the list should be sorted. That's why mKCal is sorting (so it can apply a limit with sense), and then here we sort again, because we deal with occurrences and not incidences anymore (for plain events, it's equivalent, but for recurring events not). Thus the trick in mKCal : recurring incidences are sorted first to be sure that their occurrences will be displayed if necessary, but they are not counted, to be sure that we provide at least enough incidences to create `limit` occurrences in the QML limit model.

**Review by pvuorela — APPROVED** [review]

**Pull request merged**